src/app/linux.rs

```diff
@@ -1,20 +1,449 @@
+            .into_iter()
+            .flat_map(Self::enumerate_pool)
+            .collect::<Vec<PathBuf>>();
+        Ok(SnapshotIter { idx: 0, objsets })
+    }
+
+    // Get all objset paths for a given pool.
+    // At this point all the paths should exist, because they have been
+    // checked by `required_pools_from_basepath`.  Should this somehow be
+    // expressed as an invariant?
+    fn enumerate_pool(pool: PathBuf) -> Vec<PathBuf> {
+        let is_dataset = |entry: &fs::DirEntry| {
+            let path = entry.path();
+            let name = entry.file_name();
+            let name = name.to_str().unwrap_or("");
+            path.is_file() && name.starts_with("objset")
+        };
+        fs::read_dir(pool).map_or(vec![], |dir| {
+            dir.filter_map(|e| {
+                e.map_or(None, |entry| {
+                    if is_dataset(&entry) {
+                        Some(entry.path())
+                    } else {
+                        None
+                    }
+                })
+            })
+            .collect()
+        })
+    }
+
+    fn required_pools_from_basepath(
+        basepath: &str,
+        pool: Option<&str>,
+    ) -> Result<HashSet<PathBuf>, Box<dyn Error>> {
+        let zfs_stats_path = PathBuf::from(basepath);
+        let mut pools = SnapshotIter::get_pools(zfs_stats_path.as_path())
+            .map_or_else(|err| Err(Box::new(err)), Ok)?;
+        if let Some(pool) = pool {
+            let pool: PathBuf =
+                [zfs_stats_path, PathBuf::from(pool)].iter().collect();
+            pools.retain(|p| p == &pool);
+            if pools.is_empty() {
+                let name = pool
+                    .file_name()
+                    .map(|ostr| ostr.to_str().unwrap_or(""))
+                    .unwrap_or("")
+                    .to_string();
+                return Err(Box::new(ZTopError::PoolDoesNotExist {
+                    pool: name,
+                }));
+            }
+        }
+        Ok(pools)
+    }
+
+    fn get_pools(zfs_stat_path: &Path) -> io::Result<HashSet<PathBuf>> {
+        fs::read_dir(zfs_stat_path).map(|dir| {
+            dir.filter_map(|entry| {
+                if let Ok(entry) = entry {
+                    let path = entry.path();
+                    if path.is_dir() {
+                        Some(path)
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect::<HashSet<PathBuf>>()
+        })
     }
 }
 
 impl Iterator for SnapshotIter {
     type Item = Result<Snapshot, Box<dyn Error>>;
 
     fn next(&mut self) -> Option<Self::Item> {
-        unimplemented!()
+        self.idx += 1;
+        self.objsets.get(self.idx - 1).map(|objset| {
+            fs::read_to_string(objset)
+                .map_err(|err| Box::new(err) as Box<dyn Error>)
```

Comment:
```suggestion
                .map_err(Box::new)
```
Apologies if this syntax doesn't automatically upcast to `Box<dyn Error>`.